### PR TITLE
Add recursive schemas for ClojureScript

### PR DESCRIPTION
**Pull request by viesti**
Relies on var support that was added in 0.0-2850

### CODE REVIEW

File: src/cljx/schema/core.cljx [hunk 1]
```diff
@@ -573,6 +573,26 @@
...
+    Schema
+    (walker [this]
+      (let [a (atom nil)]
+        (reset! a (start-walker
+                   (let [old subschema-walker]
+                     (clojure.core/fn [s] (if (= s this) #(@a %) (old s))))
+                   @derefable))))
+    (explain [this]
+      (let [{:keys [ns name]} (meta derefable)]
+        (list 'recursive (str ns "/" name)))))
```
Comment (w01fe) [inline src/cljx/schema/core.cljx]: can we share more of this code with the Clojure version?

File: src/cljx/schema/core.cljx [hunk 2]
```diff
@@ -573,6 +573,26 @@
...
+
+  (cljs.core/defn recursive
+  "Support for (mutually) recursive schemas by passing a var that points to a schema,
+  e.g (recursive #'ExampleRecursiveSchema)."
+  [schema]
+  (Recursive. schema)))
```
Comment (w01fe) [inline src/cljx/schema/core.cljx]: What will happen on earlier versions of cljs?  Ideally we don't break things for earlier versions.  

Can you please also add an entry to the changelog describing the addition and the new minimum cljs version?

File: test/cljx/schema/core_test.cljx [hunk 3]
```diff
@@ -263,59 +263,58 @@
...
+             :children [{:node-id 1
+                         :children [{:node-id 4}]}
+                        {:node-id "invalid-id"}]})
+
+  (valid! TestBlackNode {:black {}})
+  (valid! TestBlackNode {:black {:red {:black {}}}})
+  (invalid! TestBlackNode {:black {:black {}}})
+  (invalid! TestBlackNode {:black {:red {}}})
+
+  #+clj(let [rec (promise)
```
Comment (w01fe) [inline test/cljx/schema/core_test.cljx]: Can you put a newline before the open paren please (and below)?

**Comment by w01fe** [comment]
Looks good modulo comments, thanks for your contribution!

File: src/cljx/schema/core.cljx (hunk 2 above)
Comment (viesti) [inline src/cljx/schema/core.cljx]: Actually, now that I look a bit closer, static vars got introduced in 0.0-2496 so this should actually work with 0.0-2760. At least ```lein test``` passes.

File: src/cljx/schema/core.cljx (hunk 1 above)
Comment (viesti) [inline src/cljx/schema/core.cljx]: Good observation! I was focused on just making it work at first :). Will try to grab time to look at code sharing.

File: test/cljx/schema/core_test.cljx (hunk 3 above)
Comment (viesti) [inline test/cljx/schema/core_test.cljx]: Will do, thanks!

**Comment by w01fe** [comment]
Thanks!  LMK when you have a chance to update with the code sharing.  Or if you don't get a chance in a little while, I might take a crack at it.

**Comment by viesti** [comment]
If you have brain cycles to spare for code sharing, then go for it! I've a bit busy atm. :)

**Comment by w01fe** [comment]
Not many spare cycles but I'll get to it when I can -- if you get a chance
before then, just bump the thread to let me know and I'll get it
merged/released ASAP.  Thanks!

**Comment by viesti** [comment]
Spent the cycles from this evening :) Now the main part is shared and platform specific code is separated in a more fine grained way.

Changed also the test to use ```atom``` instead of ```promise``` to be able to share more of the test code.

**Comment by w01fe** [comment]
Thanks!  Closing in favor of #208.  I ended up squashing your changes into a single commit and making a couple small tweaks, please let me know if that's okay and I'll go ahead and merge (I preserved your authorship but probably didn't do it the "right" way).

Thanks again for your contribution!

**Comment by viesti** [comment]
Squash is ok, ended up working on many intermediate steps. Thanks! :)

**Pull request closed without merging**